### PR TITLE
NANDImporter: Check data bounds.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
Someone was very optimistic about never getting bogus data here.

### CODE REVIEW

File: Source/Core/DiscIO/NANDImporter.cpp [hunk 1]
```diff
@@ -241,7 +253,19 @@ bool NANDImporter::ExtractCertificates()
 
     const std::string pem_file_path = m_nand_root + std::string(certificate.filename);
     const ptrdiff_t certificate_offset = std::distance(content_bytes.begin(), search_result);
+    if (certificate_offset < 2 || static_cast<size_t>(certificate_offset) >= content_bytes.size())
+    {
+      ERROR_LOG_FMT(DISCIO, "ExtractCertificates: Invalid certificate offset {:#x}",
```
Comment (iwubcode) [inline Source/Core/DiscIO/NANDImporter.cpp:258]: Maybe add `content_bytes` size here as well?

File: Source/Core/DiscIO/NANDImporter.cpp [hunk 2]
```diff
@@ -241,7 +253,19 @@ bool NANDImporter::ExtractCertificates()
 
     const std::string pem_file_path = m_nand_root + std::string(certificate.filename);
     const ptrdiff_t certificate_offset = std::distance(content_bytes.begin(), search_result);
+    if (certificate_offset < 2 || static_cast<size_t>(certificate_offset) >= content_bytes.size())
```
Comment (iwubcode) [inline Source/Core/DiscIO/NANDImporter.cpp:256]:
```
const s32 min_offset = 2;
if (certificate_offset < min_offset || static_cast<size_t>(certificate_offset) >= content_bytes.size())
```

then `min_offset` can be used below.

File: Source/Core/DiscIO/NANDImporter.cpp (hunk 1 above)
Comment (iwubcode) [reply Source/Core/DiscIO/NANDImporter.cpp:258]: `ERROR_LOG_FMT(DISCIO, "ExtractCertificates: Certificate offset {:#x} must be between {} and {}", certificate_offset, min_offset, content_bytes);`

File: Source/Core/DiscIO/NANDImporter.cpp [hunk 3]
```diff
@@ -241,7 +253,19 @@ bool NANDImporter::ExtractCertificates()
...
+    {
+      ERROR_LOG_FMT(DISCIO, "ExtractCertificates: Invalid certificate offset {:#x}",
+                    certificate_offset);
+      return false;
+    }
     const u16 certificate_size = Common::swap16(&content_bytes[certificate_offset - 2]);
```
Comment (iwubcode) [inline Source/Core/DiscIO/NANDImporter.cpp:262]: `const u16 certificate_size = Common::swap16(&content_bytes[certificate_offset - min_offset ]);`

File: Source/Core/DiscIO/NANDImporter.cpp (hunk 2 above)
Comment (AdmiralCurtiss) [reply Source/Core/DiscIO/NANDImporter.cpp:256]: I'm also realizing right now that the `static_cast<size_t>(certificate_offset) >= content_bytes.size()` check is redundant, `std::search` cannot return an iterator past the end of `content_bytes` except in the not-found case, which is already checked.

Comment (iwubcode) [reply Source/Core/DiscIO/NANDImporter.cpp:256]: Oops, I completely missed that...

**Review by iwubcode — APPROVED** [review]

**Review by lioncash — APPROVED** [review]

**Pull request merged**